introduce samurai and use with CMake

neheb opened the pull request:
based on a small IRC discussion. @nbd168 suggested that it be added here. To that end, this is a newer version of https://github.com/openwrt/openwrt/pull/2874

aparcar commented on tools/samurai/Makefile:
```diff
@@ -0,0 +1,26 @@
+# 
+# Copyright (C) 2013 OpenWrt.org
```
2013? I guess this could also be added as SPDX thing.

neheb replied:
Copied from some other Makefile. I'll remove.

adschm replied:
Yes, please use SPDX identifiers on all new files.

neheb replied:
Done.

aparcar commented:
Fancy, but does this also mean all cores are always used? Sometimes it's nice to run `make -j3` to not kill my desktop UI.

neheb commented:
> Fancy, but does this also mean all cores are always used? Sometimes it's nice to run `make -j3` to not kill my desktop UI.

Interesting question. In practice, unless you're building a huge packages like domoticz or gerbera, no. So PKG_BUILD_PARALLEL has an interesting interaction with ninja. From the overall build system's point of view, ninja packages are single threaded and it treats them as such. This slows the overall queuing, but it's generally better to let ninja/samurai do its own thing.

nbd168 commented:
I don't think it's a good idea if every cmake build handles parallel building on its own. If this is the same with ninja as well, I wonder if most of the speedup isn't actually from the build system change, but from the fact that it spawns more parallel tasks. Maybe increasing the toplevel -j count could get the same amount of speedup.
Either way, this makes me skeptical if we should even use ninja or samurai for cmake.

neheb commented:
@nbd168 there's also the reliability aspect to it. Several CMake packages in the packages feed had PKG_BUILD_PARALLEL:=0 before I switched them all to ninja because the Makefiles that CMake generated could not work in parallel. Ninja does.

The manpage for samurai says:

```
-j    Run up to maxjobs jobs in parallel (default based on number of CPUs).  If zero, allow unlimited concurrent jobs.
```

It seems that the reason PKG_BUILD_PARALLEL is not doing anything is because I'm not passing $(PKG_JOBS) to the compile section.

neheb commented:
Hmm just had a thought. Maybe a middle ground could be applied where PKG_BUILD_PARALLEL=1 is added to cmake.mk as well as PKG_JOBS given that ninja/samurai doesn't suffer from the reliability problems with parallel compilation that GNU Make does. This would be closer to current behavior.

nbd168 commented:
Here's another idea: We use ninja instead of samurai, and pull in the changes from this PR: https://github.com/ninja-build/ninja/issues/1139

nbd168 committed:
build: add ninja build tool and make it available for cmake

ninja is faster at building cmake packages than make, and according to reports
also more reliable at handling parallel builds
This commit includes a patch that adds GNU make jobserver support, in order to
allow more precise control over the number of parallel tasks

Enable parallel build by default for packages using ninja

Signed-off-by: Felix Fietkau <nbd@nbd.name>
Signed-off-by: Rosen Penev <rosenp@gmail.com>

nbd168 commented:
Here's my draft commit: https://git.openwrt.org/?p=openwrt/staging/nbd.git;a=commit;h=ec0ebb17e9d0a0f1a24337986467d34bac734d56
GNU make jobserver support is working well, and I've made sure that it does not accidentally do parallel builds unless requested by the user. 
We should test all cmake packages to see if they work well with ninja. If there are no major breakages, we should probably use it by default
On larger cmake projects like mbedtls, using ninja produces a very visible improvement in build time (15s vs 20.5s), on other packages it's closer to 5% build time difference.

diizzyy commented:
https://reviews.freebsd.org/D29353 ( https://lists.alpinelinux.org/~alpine/devel/%3CBYZDB5LH4OVY.9FDVF5I3J954%40homura%3E )
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=254678
Might be of interest

neheb commented:
Looks unimplemented for samurai: https://github.com/michaelforney/samurai/issues/71

Oh well. Ninja it is.

neheb commented:
@nbd168 all packages except for 4 new ones in the packages feed are using ninja with no problems. There are no build failures on https://downloads.openwrt.org/snapshots/faillogs/ attributable to ninja.

I've tested the base feed here as well. The only package I had problems with was mt76-test because the build sections for ninja get used for the kernel stuff.

I'll do some testing with your commit. The only change I'd make is to add

```
PKG_USE_NINJA ?= 1
HOST_USE_NINJA ?= 1
```

at the top of cmake.mk

neheb committed:
package: fix cmake packages build with ninja

+= is needed for CMAKE_OPTIONS.

mt76 needs Ninja disabled as the kernel stuff uses normal make.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Respun based on @nbd168 commit. I think this should be enough to fix mt76.

neheb commented:
I just realized this commit silently introduces CMAKE_BINARY_SUBDIR support for Host packages :).

aparcar commented on package/kernel/mt76/Makefile:
```diff
@@ -13,6 +13,7 @@ PKG_SOURCE_VERSION:=22b690334c0f49b11534cc2e331c9d5e17c4a0bc
 PKG_MIRROR_HASH:=ff5e563935919d2e40c1e7254ef3bc06f7ecc5e69f8ddd12903e8f5de942d630
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
+PKG_USE_NINJA:=0
```
Why is this needed?

neheb replied:
I explained in the commit message.

mt76 is a kernel package mainly which uses make to build. The kernel stuff fails to build without this.

neheb commented:
FWIW the initial commit now includes the entire jobserver ninja PR. I did so as the later commits fixed some issues. I also added numbers to the commit message.

nbd168 committed:
build: use ninja for cmake packages by default

Speed goes from:

Executed in     178.08 secs     fish            external
usr time        20.16 mins      509.00 micros   20.16 mins
sys time        2.88 mins       39.00 micros    2.88 mins

To:

Executed in     175.90 secs     fish            external
usr time        20.19 mins      0.00 micros     20.19 mins
sys time        2.85 mins       497.00 micros   2.85 mins

Tested with "time make -j 12" on AMD Ryzen 3600

When building individual packages, the build time difference is often
significantly bigger than that.

Signed-off-by: Felix Fietkau <nbd@nbd.name>
Signed-off-by: Rosen Penev <rosenp@gmail.com>

nbd168 commented:
I've pulled in your changes into my staging tree and reworked the commits a bit, please take a look.
The change that enables ninja by default is now separate, and follows after the bugfixes.
I've also changed the PKG_BUILD_PARALLEL := 1 part to PKG_BUILD_PARALLEL ?= 1

nbd168 commented:
Another change that I've made is to remove the copy-pasted CMAKE_BINARY_SUBDIR from the host build, since the configure step didn't have the corresponding change, so this would have broken the build

neheb commented:
The ?= change seems correct. The host thing as well.

neheb commented:
I reverted the ninja-cmake.mk change in the packages feed and found a bug with this changeset.

Full explanation here:

https://github.com/openwrt/packages/pull/15012#issuecomment-808697060

```
@@ -56,7 +56,7 @@ ifeq ($(HOST_USE_NINJA),1)
   CMAKE_HOST_OPTIONS += -DCMAKE_GENERATOR="Ninja"

   define Host/Compile/Default
-       +$(NINJA) -C $(HOST_BUILD_DIR)
+       +$(NINJA) -C $(HOST_BUILD_DIR) $(1)
   endef

   define Host/Install/Default
@@ -72,7 +72,7 @@ ifeq ($(PKG_USE_NINJA),1)
   CMAKE_OPTIONS += -DCMAKE_GENERATOR="Ninja"

   define Build/Compile/Default
-       +$(NINJA) -C $(CMAKE_BINARY_DIR)
+       +$(NINJA) -C $(CMAKE_BINARY_DIR) $(1)
   endef

   define Build/Install/Default
```

is enough to fix.

neheb commented:
Updated this PR accordingly. Added missing uhttpd fix.

neheb commented:
@dhewg any thoughts?

dhewg reviewed (approved):
Nice, LGTM!
Once this is in it'll allow a few nice cleanup in the ninja package feeds

neheb commented:
I currently have https://github.com/neheb/packages/commit/8a4fd8d15181a12cd816e826fec5dbac24c4d47f and https://github.com/neheb/packages/commit/962065a1a66aadea34e72220bece515067a12454 queued up. The meson stuff should also benefit. I assume it'll be fine to remove the ninja package as well.

dhewg commented:
The package feeds provides target packages for ninja and meson too though. I don't use those myself, but I guess some do? And then it allows us to just remove the ninja host build from packages, where the target package just relies on this then.

dhewg commented:
But we could remove ninja.mk and fold the calls into meson.mk, just like cmake.mk here

neheb commented:
Would an alternative be to move ninja.mk here?

dhewg commented:
Sure, but apart from the *_BUILD_DEPENDS I think NINJA in rules.mk here covers what ninja.mk in the feeds does (not 100% certain because I haven't tried the jobserver patchset). That's more readable to me

nbd168 commented:
Changes pushed to master, thanks for testing and fixes

neheb commented:
This patchset is failing now unfortunately. tools/ninja needs to be added as a dependency of tools/cmake so that other packages can use it.
